contracts/dk/DuelistKingStaking.sol

```diff
@@ -0,0 +1,253 @@
+
+  function staking(uint256 campaignId, uint256 amountOfToken) external returns (bool) {
+    StakingCampaign memory currentCampaign = _campaignStorage[campaignId];
+    UserStakingSlot memory currentUserStakingSlot = _userStakingSlot[campaignId][msg.sender];
+    ERC20 currentToken = ERC20(currentCampaign.tokenAddress);
+
+    // User should stake at least 1 day in event time
+    require(
+      block.timestamp >= currentCampaign.startDate && block.timestamp <= currentCampaign.endDate - 1 days,
+      'DKStaking: Not in event time'
+    );
+    require(
+      currentUserStakingSlot.stakingAmountOfToken + amountOfToken <= currentCampaign.limitStakingAmountForUser,
+      'DKStaking: Token limit per user exceeded'
+    );
+
+    if (currentUserStakingSlot.stakingAmountOfToken == 0) {
+      currentUserStakingSlot.startStakingDate = uint64(block.timestamp);
+      currentUserStakingSlot.lastStakingDate = uint64(block.timestamp);
+    }
+
+    uint256 beforeBalance = currentToken.balanceOf(address(this));
+    currentToken.safeTransferFrom(msg.sender, address(this), amountOfToken);
+    uint256 afterBalance = currentToken.balanceOf(address(this));
+    require(afterBalance - beforeBalance == amountOfToken, 'DKStaking: Invalid token transfer');
+
+    currentCampaign.stakedAmountOfToken += amountOfToken;
+    require(currentCampaign.stakedAmountOfToken <= currentCampaign.maxAmountOfToken, 'DKStaking: Token limit exceeded');
+
+    currentUserStakingSlot.stakedAmountOfBoxes = estimateUserReward(currentCampaign, currentUserStakingSlot);
+    currentUserStakingSlot.lastStakingDate = uint64(block.timestamp);
+    currentUserStakingSlot.stakingAmountOfToken += amountOfToken;
+
+    _campaignStorage[campaignId] = currentCampaign;
+    _userStakingSlot[campaignId][msg.sender] = currentUserStakingSlot;
+    emit Staking(msg.sender, amountOfToken, block.timestamp);
+    return true;
+  }
+
+  function unStaking(uint256 campaignId) external returns (bool) {
+    UserStakingSlot memory currentUserStakingSlot = _userStakingSlot[campaignId][msg.sender];
+    StakingCampaign memory currentCampaign = _campaignStorage[campaignId];
+    ERC20 currentToken = ERC20(currentCampaign.tokenAddress);
+
+    require(currentUserStakingSlot.stakingAmountOfToken > 0, 'DKStaking: No token to be unstaked');
+
+    /**
+     * User unstake before endDate
+     * will be paid for penalty fee: 2% DKT and 50% reward box
+     */
+    currentUserStakingSlot.stakedAmountOfBoxes = estimateUserReward(currentCampaign, currentUserStakingSlot);
+
+    if (block.timestamp < currentCampaign.endDate) {
+      uint256 penaltyAmount = (currentUserStakingSlot.stakingAmountOfToken * 2) / 100;
+      _totalPenalty[currentCampaign.tokenAddress] += penaltyAmount;
+      currentToken.safeTransfer(msg.sender, currentUserStakingSlot.stakingAmountOfToken - penaltyAmount);
+
+      emit ClaimRewardBoxes(
+        msg.sender,
+        currentUserStakingSlot.stakedAmountOfBoxes / (2 * decimals),
```

Comment:
`decimals` Just hardcode `1000000` use the constant need you to access storage it cost an arm and a leg.